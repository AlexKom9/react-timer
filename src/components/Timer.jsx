import React, { Component } from "react";
import TimerType from "./TimerType";
import Progress from "./Progress";

class Timer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      seconds: 0,
      initialSeconds: "",
      timerIsRun: false,
      timerType: {
        hours: true,
        minutes: true,
        seconds: true
      }
    };
  }

  setSeconds = event => {
    const value = event.target.value;
    if (value < 0) return;
    this.setState({
      initialSeconds: event.target.value,
      seconds: event.target.value,
      timerIsRun: false
    });
  };

  startTimer = () => {
    this.setState(
      {
        timerIsRun: true,
        seconds: this.state.initialSeconds
      },
      () => {
        this.timer = setInterval(() => {
          this.setState({
            seconds: this.state.seconds - 1
          });
          if (this.state.seconds <= 0) {
            this.finishTimer();
          }
        }, 1000);
      }
    );
  };

  finishTimer = () => {
    this.setState(
      {
        timerIsRun: false
      },
      () => {
        clearInterval(this.timer);
      }
    );
  };

  stopTimer = () => {
    this.setState(
      {
        timerIsRun: false,
        seconds: this.state.initialSeconds
      },
      () => {
        clearInterval(this.timer);
      }
    );
  };

  getTime = () => {
    function getPercent(a, b) {
      if (Number(b) === 0) return 100;
      return 100 - Math.floor((a / b) * 100);
    }
    const { seconds, initialSeconds, timerType } = this.state;
    let time = {};
    // debugger;
    switch (JSON.stringify(timerType)) {
      default:
        // console.log("all");
        time = {
          hours: Math.floor(seconds / 3600),
          minutes: Math.floor((seconds % 3600) / 60),
          seconds: (seconds % 3600) % 60,
          hoursPercent:
            Math.floor(seconds / 3600) === 0
              ? 100
              : getPercent(seconds, initialSeconds),
          minutesPercent: getPercent(Math.floor((seconds % 3600) / 60), 60),
          secondsPercent: getPercent((seconds % 3600) % 60, 60)
        };
        break;
      case JSON.stringify({ hours: true, minutes: false, seconds: false }):
        // console.log("hours");

        time = {
          hours: Math.floor(seconds / 3600),
          minutes: false,
          seconds: false,
          hoursPercent: getPercent(seconds, initialSeconds),
          minutesPercent: false,
          secondsPercent: false
        };
        break;
      case JSON.stringify({ hours: true, minutes: true, seconds: false }):
        // console.log("hours && minutes");
        time = {
          hours: Math.floor(seconds / 3600),
          minutes: Math.floor((seconds % 3600) / 60),
          seconds: false,
          hoursPercent:
            Math.floor(seconds / 3600) === 0
              ? 100
              : getPercent(seconds, initialSeconds),
          minutesPercent: getPercent(Math.floor((seconds % 3600) / 60), 60),
          secondsPercent: false
        };
        break;
      case JSON.stringify({ hours: true, minutes: false, seconds: true }):
        // console.log("hours && seconds");
        time = {
          hours: Math.floor(seconds / 3600),
          minutes: false,
          seconds: seconds % 3600,
          hoursPercent: getPercent(
            Math.floor(seconds / 3600),
            Math.floor(initialSeconds / 3600)
          ),
          minutesPercent: false,
          secondsPercent: getPercent(seconds % 3600, initialSeconds % 3600)
        };
        break;
      case JSON.stringify({ hours: false, minutes: true, seconds: true }):
        // console.log("minutes, seconds");
        time = {
          hours: false,
          minutes: Math.floor(seconds / 60),
          seconds: seconds % 60,
          hoursPercent: false,
          minutesPercent: getPercent(
            Math.floor(seconds),
            Math.floor(initialSeconds)
          ),
          secondsPercent: getPercent(seconds % 3600, 60)
        };
        break;
      case JSON.stringify({ hours: false, minutes: true, seconds: false }):
        // console.log("minutes");
        time = {
          hours: false,
          minutes: Math.floor(seconds / 60),
          seconds: false,
          hoursPercent: false,
          minutesPercent: getPercent(seconds, initialSeconds),
          secondsPercent: false
        };
        break;
      case JSON.stringify({ hours: false, minutes: false, seconds: true }):
        // console.log("seconds");
        time = {
          hours: false,
          minutes: false,
          seconds: seconds,
          hoursPercent: false,
          minutesPercent: false,
          secondsPercent: getPercent(seconds, initialSeconds)
        };
        break;
    }
    return time;
  };

  setTimerType = timerType => {
    this.setState({
      timerType
    });
  };

  render() {
    const { timerType } = this.state;
    const time = this.getTime();
    return (
      <div className="timer">
        <div className="progress__container">
          {timerType.hours ? (
            <Progress
              name="hours"
              value={time.hours}
              percent={time.hoursPercent}
            />
          ) : null}
          {timerType.minutes ? (
              <Progress
                name="minutes"
                value={time.minutes}
                percent={time.minutesPercent}
              />
          ) : null}
          {timerType.seconds ? (
            <Progress
              name="seconds"
              value={time.seconds}
              percent={time.secondsPercent}
            />
          ) : null}
        </div>

        <div className="timer__start">
          <input
            id="timer-value"
            disabled={this.state.timerIsRun}
            type="number"
            value={this.state.initialSeconds}
            onChange={this.setSeconds}
          />
          <button
            id="start-timer"
            disabled={
              this.state.timerIsRun || !Number(this.state.initialSeconds)
            }
            onClick={this.startTimer}
          >
            Start Timer
          </button>
          <button
            id="stop-timer"
            disabled={!this.state.timerIsRun}
            onClick={this.stopTimer}>
            Stop Timer
          </button>
        </div>

        <div className="timer__type">
          <TimerType setTimerType={this.setTimerType} />
        </div>
      </div>
    );
  }
}

Timer.propTypes = {};

export default Timer;
