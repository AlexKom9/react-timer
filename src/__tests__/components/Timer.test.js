import React from "react";
import Timer from "../../components/Timer";

import { shallow, mount } from "enzyme";

describe(`Timer`, () => {
  test(`initial render`, () => {
    const wrapper = shallow(<Timer />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("initial progress amount", () => {
    const wrapper = mount(<Timer />);
    expect(wrapper.find(".progress__container").find(".progress")).toHaveLength(
      3
    );
  });

  test("click Start Timer with input 10 seconds", () => {
    const wrapper = mount(<Timer />);

    wrapper.find("#timer-value").simulate("change", {
      target: {
        value: 10
      }
    });

    wrapper.find("#start-timer").simulate("click");
    console.log(wrapper.state());
    expect(wrapper.state("seconds")).toBe(10);
    expect(wrapper.state("initialSeconds")).toBe(10);
    expect(wrapper.state("timerIsRun")).toBe(true);
    expect(wrapper.find("#timer-value").prop('disabled')).toBe(true);
    expect(wrapper.find("#start-timer").prop('disabled')).toBe(true);
    expect(wrapper.find("#stop-timer").prop('disabled')).toBe(false);
  });
});
