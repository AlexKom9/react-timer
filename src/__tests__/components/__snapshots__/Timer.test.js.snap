// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer initial render 1`] = `
"<div class=\\"timer\\"><div class=\\"progress__container\\"><div class=\\"progress\\"><svg width=\\"120\\" height=\\"120\\" viewBox=\\"0 0 120 120\\"><defs><filter id=\\"inset-shadow\\"><feFlood flood-color=\\"black\\"></feFlood><feComposite operator=\\"xor\\" in2=\\"SourceGraphic\\"></feComposite><feGaussianBlur stdDeviation=\\"2\\"></feGaussianBlur><feComposite operator=\\"in\\" in2=\\"SourceGraphic\\" result=\\"map\\"></feComposite><feDiffuseLighting lighting-color=\\"#fff\\" surfaceScale=\\"4\\" diffuseConstant=\\"1.2\\"><feSpotLight x=\\"0\\" y=\\"-30\\" z=\\"230\\"></feSpotLight></feDiffuseLighting><feBlend mode=\\"multiply\\" in=\\"SourceGraphic\\"></feBlend><feComposite operator=\\"in\\" in2=\\"SourceGraphic\\"></feComposite></filter><pattern id=\\"bgi\\" patternUnits=\\"userSpaceOnUse\\" width=\\"32\\" height=\\"32\\"><image xlink:href=\\"bgi.jpg\\" width=\\"2\\" height=\\"2\\"></image></pattern></defs><circle class=\\"circle__background\\" filter=\\"url(#inset-shadow)\\" cx=\\"60\\" cy=\\"60\\" r=\\"56\\" stroke-width=\\"8px\\" fill=\\"none\\"></circle><circle class=\\"circle__progress hours\\" cx=\\"60\\" cy=\\"60\\" r=\\"56\\" stroke-width=\\"8px\\" fill=\\"transparent\\" transform=\\"rotate(-90 60 60)\\" style=\\"stroke-dasharray:351.85837720205683;stroke-dashoffset:5.684341886080802e-14\\"></circle><foreignObject class=\\"node\\" width=\\"120\\" height=\\"120\\"><div><span class=\\"circle__value hours\\">0</span><p class=\\"circle__name hours\\">hours</p></div></foreignObject></svg></div><div class=\\"progress\\"><svg width=\\"120\\" height=\\"120\\" viewBox=\\"0 0 120 120\\"><defs><filter id=\\"inset-shadow\\"><feFlood flood-color=\\"black\\"></feFlood><feComposite operator=\\"xor\\" in2=\\"SourceGraphic\\"></feComposite><feGaussianBlur stdDeviation=\\"2\\"></feGaussianBlur><feComposite operator=\\"in\\" in2=\\"SourceGraphic\\" result=\\"map\\"></feComposite><feDiffuseLighting lighting-color=\\"#fff\\" surfaceScale=\\"4\\" diffuseConstant=\\"1.2\\"><feSpotLight x=\\"0\\" y=\\"-30\\" z=\\"230\\"></feSpotLight></feDiffuseLighting><feBlend mode=\\"multiply\\" in=\\"SourceGraphic\\"></feBlend><feComposite operator=\\"in\\" in2=\\"SourceGraphic\\"></feComposite></filter><pattern id=\\"bgi\\" patternUnits=\\"userSpaceOnUse\\" width=\\"32\\" height=\\"32\\"><image xlink:href=\\"bgi.jpg\\" width=\\"2\\" height=\\"2\\"></image></pattern></defs><circle class=\\"circle__background\\" filter=\\"url(#inset-shadow)\\" cx=\\"60\\" cy=\\"60\\" r=\\"56\\" stroke-width=\\"8px\\" fill=\\"none\\"></circle><circle class=\\"circle__progress minutes\\" cx=\\"60\\" cy=\\"60\\" r=\\"56\\" stroke-width=\\"8px\\" fill=\\"transparent\\" transform=\\"rotate(-90 60 60)\\" style=\\"stroke-dasharray:351.85837720205683;stroke-dashoffset:5.684341886080802e-14\\"></circle><foreignObject class=\\"node\\" width=\\"120\\" height=\\"120\\"><div><span class=\\"circle__value minutes\\">0</span><p class=\\"circle__name minutes\\">minutes</p></div></foreignObject></svg></div><div class=\\"progress\\"><svg width=\\"120\\" height=\\"120\\" viewBox=\\"0 0 120 120\\"><defs><filter id=\\"inset-shadow\\"><feFlood flood-color=\\"black\\"></feFlood><feComposite operator=\\"xor\\" in2=\\"SourceGraphic\\"></feComposite><feGaussianBlur stdDeviation=\\"2\\"></feGaussianBlur><feComposite operator=\\"in\\" in2=\\"SourceGraphic\\" result=\\"map\\"></feComposite><feDiffuseLighting lighting-color=\\"#fff\\" surfaceScale=\\"4\\" diffuseConstant=\\"1.2\\"><feSpotLight x=\\"0\\" y=\\"-30\\" z=\\"230\\"></feSpotLight></feDiffuseLighting><feBlend mode=\\"multiply\\" in=\\"SourceGraphic\\"></feBlend><feComposite operator=\\"in\\" in2=\\"SourceGraphic\\"></feComposite></filter><pattern id=\\"bgi\\" patternUnits=\\"userSpaceOnUse\\" width=\\"32\\" height=\\"32\\"><image xlink:href=\\"bgi.jpg\\" width=\\"2\\" height=\\"2\\"></image></pattern></defs><circle class=\\"circle__background\\" filter=\\"url(#inset-shadow)\\" cx=\\"60\\" cy=\\"60\\" r=\\"56\\" stroke-width=\\"8px\\" fill=\\"none\\"></circle><circle class=\\"circle__progress seconds\\" cx=\\"60\\" cy=\\"60\\" r=\\"56\\" stroke-width=\\"8px\\" fill=\\"transparent\\" transform=\\"rotate(-90 60 60)\\" style=\\"stroke-dasharray:351.85837720205683;stroke-dashoffset:5.684341886080802e-14\\"></circle><foreignObject class=\\"node\\" width=\\"120\\" height=\\"120\\"><div><span class=\\"circle__value seconds\\">0</span><p class=\\"circle__name seconds\\">seconds</p></div></foreignObject></svg></div></div><div class=\\"timer__start\\"><input type=\\"number\\" id=\\"timer-value\\" value=\\"\\"/><button id=\\"start-timer\\" disabled=\\"\\">Start Timer</button><button id=\\"stop-timer\\" disabled=\\"\\">Stop Timer</button></div><div class=\\"timer__type\\"><div class=\\"timer-type\\"><input type=\\"text\\" class=\\"timer-type__input\\" value=\\"hours, minutes, seconds\\"/><br/><textarea class=\\"timer-type__text\\" cols=\\"30\\" rows=\\"10\\">hours
minutes
seconds</textarea><br/><select multiple=\\"\\" class=\\"timer-type__select\\"><option value=\\"hours\\" selected=\\"\\">hours</option><option value=\\"minutes\\" selected=\\"\\">minutes</option><option value=\\"seconds\\" selected=\\"\\">seconds</option></select></div></div></div>"
`;
